File: c_cpp/Tree/Binary_Tree.cpp
#include "binarytree.h"

BinaryTree::~BinaryTree() {
    function<void(TreeNode*)> f = [&](TreeNode* node) {
        if (!node) return;
        cout << node->val << "deleted.." << endl;
        TreeNode* Left_Tree = node->left;
        TreeNode* Right_Tree = node->right;
        delete node;
        if (Left_Tree) f(Left_Tree);
        if (Right_Tree) f(Right_Tree);
    };
    // 前序析构
    function<void(TreeNode*)> g = [&](TreeNode* node) {
        if (!node) return;
        stack<TreeNode*> st;
        st.push(node);
        while (!st.empty()) {
            node = st.top();
            st.pop();
            cout << node->val << "deleted...\n";
            if (node->right) st.push(node->right);
            if (node->left) st.push(node->left);
            delete node;
        }
    };
    // recur dtor by preorder
    f(root);
    // iter dtor by preorder
    /* g(root); */
}

// 递归版本生成二叉树(需要捕获用户输入)
void BinaryTree::add_recur1() {
    function<TreeNode*(void)> f = [&](void) {
        int data;
        cin >> data;
        if (data == 0) { return (TreeNode*)nullptr; }
        TreeNode* node = new TreeNode(data);
        node->val = data;
        node->left = f();
        node->right = f();
        return node;
    };
    root = f();
}

// 递归生成二叉树: 直接遍历外部数组实现
void BinaryTree::add_recur2(vector<int>& item) {
    function<TreeNode*()> f = [&]() {
        if (item.empty() || item.front() == 0) {
            item.erase(item.begin());
            return (TreeNode*)nullptr;
        }
        TreeNode* node = new TreeNode(item.front());
        node->val = item.front();
        item.erase(item.begin());
        node->left = f();
        node->right = f();
        return node;
    };
    root = f();
}


void BinaryTree::add_iter(int item) {
    TreeNode* node = new TreeNode(item);

    if (!root) {
        root = node;
        return;
    }
    // 用队列模拟二叉树的层序遍历
    queue<TreeNode*> que;
    que.push(root);
    while (!que.empty()) {
        TreeNode* cur_node = que.front();
        que.pop();

        if (!cur_node->left) {
            cur_node->left = node;
            return;
        } else
            que.push(cur_node->left);

        // 接下来同样判断右边
        if (!cur_node->right) {
            cur_node->right = node;
            return;
        } else
            que.push(cur_node->right);
    }
}

void BinaryTree::breadth_travel() {
    queue<TreeNode*> que;
    que.push(root);
    vector<int> ret;
    if (!root) return;
    while (!que.empty()) {
        TreeNode* cur_node = que.front();
        que.pop();
        // 也可以直接打印, 省去了定义数组的步骤
        // cout<< cur_node->val<<endl;
        ret.push_back(cur_node->val);

        if (cur_node->left) que.push(cur_node->left);
        if (cur_node->right) que.push(cur_node->right);
    }
    cout << ret << endl;
}

/********** pre order **********/

void BinaryTree::pre_order() {
    vector<int> ret;
    auto recur_0 = [&ret](auto&& self, TreeNode* node) {
        // 递归终止条件
        if (!node) return;
        ret.push_back(node->val);
        self(self, node->left);
        self(self, node->right);
    };
    recur_0(recur_0, root);
    cout << ret << endl;
}

void BinaryTree::pre_order1() {
    vector<int> ret;
    stack<TreeNode*> st;
    if (root) st.push(root);
    while (!st.empty()) {
        auto node = st.top();
        st.pop();
        ret.push_back(node->val);
        if (node->right) st.push(node->right);
        if (node->left) st.push(node->left);
    }

    cout << ret << endl;
}
void BinaryTree::pre_order2() {
    vector<int> ret{};
    stack<pair<int, TreeNode*>> st;
    if (root) st.push(make_pair(0, root));
    /* int color; */
    /* TreeNode* node; */
    while (!st.empty()) {
        auto [color, node] = st.top();
        st.pop();
        if (node == nullptr) continue;
        if (color == 0) {
            st.push(make_pair(0, node->right));
            st.push(make_pair(0, node->left));
            st.push(make_pair(1, node));
        } else {
            ret.push_back(node->val);
        }
    }
    cout << ret << endl;
}

void BinaryTree::morris_preorder() {
    if (!root) return;
    vector<int> ans{};
    TreeNode* cur = root;
    TreeNode* mostRight;
    while (cur) {
        mostRight = cur->left;
        if (mostRight) {
            while (mostRight->right and mostRight->right != cur)
                mostRight = mostRight->right;
            if (!mostRight->right) {
                mostRight->right = cur;
                ans.push_back(cur->val);
                cur = cur->left;
                continue;
            } else
                mostRight->right = nullptr;
        } else
            ans.push_back(cur->val);
        cur = cur->right;
    }
    cout << ans << endl;
}

// 一般方法(前中后都可用, 通过空节点'标记'根节点位置)
void BinaryTree::pre_order3() {
    vector<int> ret;
    stack<TreeNode*> st;
    if (root) st.push(root);
    while (!st.empty()) {
        auto node = st.top();
        st.pop();
        if (node) {
            // 添加右节点, 空节点不入栈
            if (node->right) st.push(node->right);
            // 添加左节点, 空节点不入栈
            if (node->left) st.push(node->left);
            st.push(node); // 中(根)节点标记
            st.push(nullptr);
        } else {
            node = st.top();
            st.pop();
            ret.push_back(node->val);
        }
    }
    cout << ret << endl;
}


/*Morris 前序遍历的流程如下：
第一步.当前结点的左孩子是否为空，若是则输出当前结点，并更新当前结点为当前结点的右孩子；否则进入第二步.
第二步.在当前结点的左子树中寻找左子树中最右结点作为前驱结点
a.若前驱结点的右孩子为空,则将前驱结点的右孩子指向当前结点,输出当前结点.当前结点更新为当前结点的左孩子,
进入第三步.
b.若前驱结点的右孩子为当前结点（不为空）,将前驱结点的右孩子置NULL,当前结点更新为当前结点的右孩子,
进入第三步.
第三步.若当前结点不为空,进入第一步；否则程序结束
*/
void BinaryTree::morris_preorder1() {
    if (!root) return;
    vector<int> ans{};
    TreeNode *cur = root, *pre = nullptr;
    while (cur) {
        if (!cur->left) {
            ans.push_back(cur->val);
            cur = cur->right;
        } else {
            // 定位左子树
            pre = cur->left;
            // 找左子树的最右节点(不能与cur相同)
            while (pre->right && pre->right != cur) pre = pre->right;
            if (!pre->right) {
                pre->right = cur;
                ans.push_back(cur->val);
                cur = cur->left;
            } else {
                pre->right = nullptr;
                cur = cur->right;
            }
        }
    }
    cout << ans << endl;
}

/********** in order **********/

void BinaryTree::in_order() {
    vector<int> ret;
    auto recur_1 = [&ret](auto&& self, TreeNode* node) {
        if (!node) return;

        self(self, node->left);
        ret.push_back(node->val);
        self(self, node->right);
    };
    recur_1(recur_1, root);
    cout << ret << endl;
}

void BinaryTree::in_order1() {
    vector<int> ret;
    stack<TreeNode*> st;
    auto cur = root;
    while (!st.empty() || cur) {
        if (cur) {
            st.push(cur);
            cur = cur->left; // 左
        } else {
            cur = st.top();
            st.pop();
            ret.push_back(cur->val); // 根
            cur = cur->right;        // 右
        }
    }
    cout << ret << endl;
}

// 一般方法(前中后都可用, 通过空节点'标记'根节点位置)
void BinaryTree::in_order2() {
    vector<int> ret;
    stack<TreeNode*> st;
    if (root) st.push(root);
    while (!st.empty()) {
        auto node = st.top();
        st.pop();
        if (node) {
            // 添加右节点, 空节点不入栈
            if (node->right) st.push(node->right);
            st.push(node); // 中(根)节点标记
            st.push(nullptr);
            // 添加左节点, 空节点不入栈
            if (node->left) st.push(node->left);
        } else {
            node = st.top();
            st.pop();
            ret.push_back(node->val);
        }
    }
    cout << ret << endl;
}

void BinaryTree::in_order3() {
    vector<int> ret{};
    stack<pair<int, TreeNode*>> st;
    if (root) st.push(make_pair(0, root));
    /* int color; */
    /* TreeNode* node; */
    while (!st.empty()) {
        auto [color, node] = st.top(); // C++17
        st.pop();
        if (node == nullptr) continue;
        if (color == 0) {
            st.push(make_pair(0, node->right));
            st.push(make_pair(1, node));
            st.push(make_pair(0, node->left));
        } else {
            ret.push_back(node->val);
        }
    }
    cout << ret << endl;
}

void BinaryTree::morris_inorder() {
    if (!root) return;
    vector<int> ans{};
    TreeNode* cur = root;
    TreeNode* mostRight;
    while (cur) {
        mostRight = cur->left;
        if (mostRight) {
            while (mostRight->right and mostRight->right != cur)
                mostRight = mostRight->right;
            if (!mostRight->right) {
                mostRight->right = cur;
                cur = cur->left;
                continue;
            } else
                mostRight->right = nullptr;
        }
        ans.push_back(cur->val);
        cur = cur->right;
    }
    cout << ans << endl;
}

/*Morris 中序遍历的流程如下:
第一步.当前结点的左孩子是否为空,若是则输出当前结点,并更新当前结点为当前结点的右孩子;否则进入第二步.
第二步.在当前结点的左子树中寻找左子树中最右结点作为前驱结点
a.若前驱结点的右孩子为空,则将前驱结点的右孩子指向当前结点,当前结点更新为当前结点的左孩子;进入第三步;
b.若前驱结点的右孩子为当前结点（不为空）,将前驱结点的右孩子置NULL,输出当前结点,当前结点更新为当前结点的右孩子,进入第三步;
第三步.若当前结点不为空, 进入第一步; 否则程序结束*/
void BinaryTree::morris_inorder1() {
    if (!root) return;
    vector<int> ans{};
    TreeNode *cur = root, *pre = nullptr;
    while (cur) {
        if (!cur->left) {
            ans.push_back(cur->val);
            cur = cur->right;
        } else {
            pre = cur->left;
            while (pre->right && pre->right != cur) pre = pre->right;
            if (!pre->right) {
                pre->right = cur;
                cur = cur->left;
            } else {
                pre->right = nullptr;
                ans.push_back(cur->val);
                cur = cur->right;
            }
        }
    }
    cout << ans << endl;
}

/********** post order **********/

void BinaryTree::post_order() {
    vector<int> ret;
    auto recur_2 = [&ret](auto&& self, TreeNode* node) {
        if (!node) return;

        self(self, node->left);
        self(self, node->right);
        ret.push_back(node->val);
    };
    recur_2(recur_2, root);
    cout << ret << endl;
}

// 从前序的逆序得到
void BinaryTree::post_order1() {
    stack<TreeNode*> st;
    vector<int> ret;
    if (root) { st.push(root); }
    while (!st.empty()) {
        auto node = st.top();
        st.pop();
        ret.push_back(node->val);
        if (node->left) st.push(node->left);
        if (node->right) st.push(node->right);
    }
    // 进行逆序操作: 或者采用逆序算法
    // vector<int> ret1;
    // for (auto it = ret.rbegin(); it != ret.rend(); it++)
    // ret1.push_back(*it);
    reverse(ret.begin(), ret.end());

    cout << ret << endl;
}

void BinaryTree::post_order2() {
    stack<TreeNode*> st;
    vector<int> ret;
    auto node = root;
    while (!st.empty() || node) {
        while (node) { // 找最左边节点的最右边节点(即遍历的开始)
            st.push(node);
            node = node->left ? node->left : node->right;
        }
        node = st.top();
        st.pop();
        ret.push_back(node->val);
        node = // 难想
            !st.empty() && st.top()->left == node ? st.top()->right : nullptr;
    }

    cout << ret << endl;
}

void BinaryTree::post_order3() {
    stack<TreeNode*> st;
    vector<int> ret;
    TreeNode* node = root;
    while (!st.empty() || node) {
        if (node) {
            ret.push_back(node->val);
            if (node->left) st.push(node->left);
            node = node->right;
        } else {
            node = st.top();
            st.pop();
        }
    }
    reverse(ret.begin(), ret.end());

    cout << ret << endl;
}

void BinaryTree::post_order4() {
    vector<int> ret{};
    stack<pair<int, TreeNode*>> st;
    if (root) st.push(make_pair(0, root));
    /* int color; */
    /* TreeNode* node; */
    while (!st.empty()) {
        auto [color, node] = st.top();
        st.pop();
        if (node == nullptr) continue;
        if (color == 0) {
            st.push(make_pair(1, node));
            st.push(make_pair(0, node->right));
            st.push(make_pair(0, node->left));
        } else {
            ret.push_back(node->val);
        }
    }
    cout << ret << endl;
}

// 一般方法(前中后都可用, 通过空节点'标记'根节点位置)
void BinaryTree::post_order5() {
    vector<int> ret;
    stack<TreeNode*> st;
    if (root) st.push(root);
    while (!st.empty()) {
        auto node = st.top();
        st.pop();
        if (node) {
            st.push(node); // 中(根)节点标记
            st.push(nullptr);
            // 添加右节点, 空节点不入栈
            if (node->right) st.push(node->right);
            // 添加左节点, 空节点不入栈
            if (node->left) st.push(node->left);
        } else {
            node = st.top();
            st.pop();
            ret.push_back(node->val);
        }
    }
    cout << ret << endl;
}


/*Morris 后序遍历的流程如下:
新建一个Dummy结点,该结点的左孩子指向树根root,将Dummy作为当前结点;
第一步.当前结点的左孩子是否为空,更新当前结点为当前结点的右孩子;否则进入第二步;
第二步.在当前结点的左子树中寻找中序遍历下的前驱结点（左子树中最右结点）:
a.若前驱结点的右孩子为空,则将前驱结点的右孩子指向当前结点,当前结点更新为当前结点的左孩子,进入3;
b.若前驱结点的右孩子为当前结点（不为空）,反转当前结点左孩子到前驱结点之间的路径,输出该路径所有结点;再反转恢复原状.
将前驱结点的右孩子置NULL,当前结点更新为当前结点的右孩子,进入第三步;
第三步.若当前结点不为空,进入第一步;否则程序结束*/
void BinaryTree::morris_postorder() {
    if (!root) return;
    vector<int> ans{};

    function<TreeNode*(TreeNode*)> reverseEdge = [&](TreeNode* node) {
        TreeNode *pre = nullptr, *next = nullptr;
        while (node) {
            next = node->right;
            node->right = pre;
            pre = node;
            node = next;
        }
        return pre;
    };

    function<void(TreeNode*)> printEdge = [&](TreeNode* node) {
        TreeNode* tail = reverseEdge(node);
        TreeNode* cur = tail;
        while (cur) {
            ans.push_back(cur->val);
            cur = cur->right;
        }
        reverseEdge(tail);
    };
    TreeNode *cur = root, *mostRight;
    while (cur) {
        mostRight = cur->left;
        if (mostRight) {
            while (mostRight->right and mostRight->right != cur)
                mostRight = mostRight->right;
            if (!mostRight->right) {
                mostRight->right = cur;
                cur = cur->left;
                continue;
            } else {
                mostRight->right = nullptr;
                printEdge(cur->left);
            }
        }
        cur = cur->right;
    }
    printEdge(root);
    cout << ans << endl;
}

/*
1.新建临时节点cur,令该节点为 root;
2.如果当前节点的左子节点为空,则遍历当前节点的右子节点;
3.如果当前节点的左子节点不为空,在当前节点的左子树中找到当前节点在中序遍历下的前驱节点;
    a.如果前驱节点的右子节点为空,将前驱节点的右子节点设置为当前节点,当前节点更新为当前节点的左子节点.
    b.如果前驱节点的右子节点为当前节点,将它的右子节点重新设为空.
倒序输出从当前节点的左子节点到该前驱节点这条路径上的所有节点.
当前节点更新为当前节点的右子节点.
4.重复步骤2和步骤3,直到遍历结束.
*/
void BinaryTree::morris_postorder1() {
    if (!root) return;
    vector<int> res;

    function<void(TreeNode*)> addPath = [&](TreeNode* node) {
        // addPath 添加至结果数组, 然后完成部分数组反转
        int count{};
        while (node) {
            ++count;
            res.emplace_back(node->val);
            node = node->right;
        }
        reverse(res.end() - count, res.end());
    };
    TreeNode *cur = root, *pre = nullptr;
    while (cur) {
        pre = cur->left;
        if (!pre)
            cur = cur->right;
        else {
            while (pre->right && pre->right != cur) pre = pre->right;
            if (!pre->right) {
                pre->right = cur;
                cur = cur->left;
            } else {
                pre->right = nullptr;
                addPath(cur->left);
                cur = cur->right;
            }
        }
    }
    addPath(root);
    cout << res << endl;
}

int main(int argc, char const* argv[]) {
    BinaryTree tree;
    // 第一种生成方式, 层序生成(使用队列进行循环)
    // for (int i = 1; i < 8; ++i) {
    //     tree.add_iter(i);
    // }

    // // 第二种生成方式, 递归生成(前序递归), 需要传入用户输入
    // tree.add_recur1();

    // // 第三种(同第二种), 递归生成, 直接读取数组即可
    // vector<int> arr = {1, 2, 0, 4, 0, 0, 3, 0, 5, 0, 0};
    // vector<int> arr = {1, 2, 3, 4, 5, 6, 7, 0, 0, 0, 0, 0, 0, 0, 0};
    // vector<int> arr = {1, 2, 4, 0, 0, 5, 0, 0, 3, 6, 0, 0, 7, 0, 0};
    // vector<int> arr = {1, 2, 0, 5, 0, 0, 3, 0, 0};
    vector<int> arr = {1, 2, 4, 0, 5, 0, 0, 0, 3, 0, 0};
    /*
        1
       / \
      2   3
     /
    4
     \
      5
    */
    tree.add_recur2(arr);

    // 下面是遍历
    // 广度遍历
    cout << "breadth_travel:\n";
    tree.breadth_travel();

    // 前序遍历(递归)
    cout << "\npre_order_recur:\n";
    tree.pre_order();
    // 前序遍历(循环)
    cout << "pre_order_iter1:\n";
    tree.pre_order1();
    cout << "pre_order_iter2:\n";
    tree.pre_order2();
    cout << "pre_order_iter3:\n";
    tree.pre_order3();
    cout << "morris_preorder:\n";
    tree.morris_preorder();
    tree.morris_preorder1();


    // 中序遍历(递归)
    cout << "\nin_order_recur:\n";
    tree.in_order();
    // 中序遍历(循环)
    cout << "in_order_iter1:\n";
    tree.in_order1();
    cout << "in_order_iter2:\n";
    tree.in_order2();
    cout << "in_order_iter3:\n";
    tree.in_order3();
    cout << "morris_inorder:\n";
    tree.morris_inorder();
    tree.morris_inorder1();


    // 后序遍历(递归)
    cout << "\npost_order_recur:\n";
    tree.post_order();
    // 后序遍历(循环)
    cout << "post_order_iter1:\n";
    tree.post_order1();
    cout << "post_order_iter2:\n";
    tree.post_order2();
    cout << "post_order_iter3:\n";
    tree.post_order3();
    cout << "post_order_iter4:\n";
    tree.post_order4();
    cout << "post_order_iter5:\n";
    tree.post_order5();
    cout << "morris_postorder:\n";
    tree.morris_postorder();
    tree.morris_postorder1();

    return 0;
    /*
    breadth_travel:
    [1, 2, 3, 4, 5, 6, 7]
    pre_order_recur:
    [1, 2, 4, 5, 3, 6, 7]
    in_order_recur:
    [4, 2, 5, 1, 6, 3, 7]
    post_order_recur:
    [4, 5, 2, 6, 7, 3, 1]
    */
}

File: py/data_structure/binary_Tree.py
from collections import deque

# 首先定义树的根节点


class Node(object):
    """docstring for Node"""

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# 定义一棵二叉树
class BinaryTree(object):
    """docstring for BinaryTree"""

    def __init__(self):
        # 根节点
        self.root = None

    def add(self, item):
        node = Node(item)
        # 向树中插入元素, 使用队列存储元素, 读取与弹出
        if not self.root:
            self.root = node
            return
        # 用顺序表实现队列, 先入先出FIFO
        queue = [self.root]

        while queue:
            cur_node = queue.pop(0)
            if cur_node.val is None:
                continue
            # 若当前节点的左孩子为空, 将节点赋给当前节点左孩子
            if not cur_node.left:
                cur_node.left = node
                return
            else:
                queue.append(cur_node.left)
            if not cur_node.right:
                cur_node.right = node
                return
            else:
                queue.append(cur_node.right)

    def breadth_travel(self):
        """广度遍历: 方法同add, 是一种反过来的操作"""
        # 使用队列
        queue = [self.root]
        ret = []
        if self.root is None:
            return
        while queue:
            cur_node = queue.pop(0)
            # 打印结点值
            # print(cur_node.val, end=" ")
            ret.append(cur_node.val)
            if cur_node.left:
                queue.append(cur_node.left)
            if cur_node.right:
                queue.append(cur_node.right)
        print(ret)

    def breadth_travel1(self):
        """广度遍历: 方法同add, 是一种反过来的操作"""
        # 使用队列
        q = [self.root]
        if self.root is None:
            return
        level = 1
        while q:
            nq = []
            for cur_node in q:
                if cur_node.left:
                    nq.append(cur_node.left)
                if cur_node.right:
                    nq.append(cur_node.right)
            print(f"第{level}层元素: {[i.val for i in q]}")
            q = nq
            level += 1

    def breadth_travel2(self):
        """广度遍历: 方法同add, 是一种反过来的操作"""
        # 使用队列
        q = deque([self.root])
        if self.root is None:
            return
        level = 1
        while q:
            print(f"第{level}层元素: {[i.val for i in q]}")
            for _ in range(len(q)):
                cur_node = q.popleft()
                if cur_node.left:
                    q.append(cur_node.left)
                if cur_node.right:
                    q.append(cur_node.right)
            level += 1

    def pre_order(self):
        """前序遍历: 递归方法"""
        # if node is None:
        #     return
        # print(node.val, end=" ")
        # self.pre_order(node.left)
        # self.pre_order(node.right)
        ret = []

        def recur_0(node):
            if node:
                ret.append(node.val)
                recur_0(node.left)
                recur_0(node.right)

        recur_0(self.root)
        print(ret)

    def in_order(self):
        """中序遍历"""
        # if node is None:
        #     return
        # self.in_order(node.left)
        # print(node.val, end=" ")
        # self.in_order(node.right)
        ret = []

        def recur_1(node):
            if node:
                recur_1(node.left)
                ret.append(node.val)
                recur_1(node.right)

        recur_1(self.root)
        print(ret)

    def post_order(self):
        """后序遍历"""
        # 递归遍历跳出的条件
        # if node is None:
        #     return
        # self.pre_order(node.left)
        # self.pre_order(node.right)
        # print(node.val, end=" ")
        ret = []

        def recur_2(node):
            if node:
                recur_2(node.left)
                recur_2(node.right)
                ret.append(node.val)

        recur_2(self.root)
        print(ret)

    def pre_order_1(self):
        """前序遍历(中左右): 非递归, 需要使用栈(递归的本质是栈实现)来实现"""
        # 定义一个栈
        st = []
        # 定义顺序数组
        result_arr = []
        if self.root:
            st.append(self.root)
        while st:
            node = st.pop()
            # 中
            result_arr.append(node.val)
            # 右
            if node.right:
                st.append(node.right)
            # 左
            if node.left:
                st.append(node.left)
        print(result_arr)

    def in_order_1(self):
        """中序遍历(左中右), 需要指针(由于遍历的节点顺序和处理的节点顺序不同)"""
        st = []
        result_arr = []
        cur_node = self.root  # 这里用指针是防止根节点被改变影响之后的操作
        while cur_node or st:
            if cur_node:
                # 利用指针访问结点,访问到最底层数据
                # 结点入栈
                st.append(cur_node)
                # 左
                cur_node = cur_node.left
            else:
                cur_node = st.pop()
                # 中
                result_arr.append(cur_node.val)
                # 右
                cur_node = cur_node.right
        print(result_arr)

    def in_order_2(self):
        """中序遍历(左中右), 通解"""
        st = []
        result_arr = []
        if self.root:
            st.append(self.root)
        while st:
            node = st.pop()
            if node:
                if node.right:
                    st.append(node.right)
                st.append(node)
                # 空节点入栈作为标记
                st.append(None)
                if node.left:
                    st.append(node.left)
            else:
                # 若为空节点,直接捕获下一个节点
                node = st.pop()
                result_arr.append(node.val)
        print(result_arr)

    def in_order_3(self):
        WHITE, GRAY = 0, 1
        res = []
        stack = [(WHITE, self.root)]
        while stack:
            color, node = stack.pop()
            if node is None:
                continue
            if color == WHITE:
                stack.append((WHITE, node.right))
                stack.append((GRAY, node))
                stack.append((WHITE, node.left))
            else:
                res.append(node.val)
        print(res)

    def in_order_4(self):
        res = []
        stack = [self.root]
        while stack:
            node = stack.pop()
            if isinstance(node, Node):
                stack.extend([node.right, node.val, node.left])
            elif isinstance(node, int):
                res.append(node)
        print(res)

    def post_order_1(self):
        """后序遍历(左右中),直接由前序遍历得到"""
        # 定义一个栈
        st = []
        # 定义顺序数组
        result_arr = []
        if self.root:
            st.append(self.root)
        while st:
            node = st.pop()
            # 中
            result_arr.append(node.val)
            # 左
            if node.left:
                st.append(node.left)
            # 右
            if node.right:
                st.append(node.right)
        print(result_arr[::-1])

    def post_order_2(self):
        """后序遍历(左右中)"""
        # 定义一个栈
        st = []
        # 定义顺序数组
        result_arr = []
        node = self.root
        while st or node:
            while node:
                st.append(node)
                # 遍历二叉树直到结点不再含有左节点(右节点)
                node = node.left if node.left else node.right
            node = st.pop()
            # 最后加入中结点
            result_arr.append(node.val)
            # 判断并开始遍历右节点(node指向右节点), 然后继续进行入栈操作(while内循环)
            node = st[-1].right if st and st[-1].left == node else None
        print(result_arr)

    def print_Tree(self):
        # m:层数
        # n:列数
        q = deque([self.root])
        m = 0
        while q:
            for _ in range(len(q)):
                cur = q.popleft()
                # if not (cur.left or cur.right): continue
                if cur.left:
                    q.append(cur.left)
                if cur.right:
                    q.append(cur.right)
            m += 1
        n = (1 << m) - 1
        ans = [[" "] * n for _ in range(m)]
        branch = [[" "] * n for _ in range(m)]
        bq = deque([[0, (n - 1) // 2, self.root, ""]])
        while bq:
            for _ in range(len(bq)):
                r, c, cur, slash = bq.popleft()
                if cur.val is None:
                    continue
                ans[r][c] = str(cur.val)
                # 叶结点不能对斜杠/反斜杠进行移动
                if r == m - 1:
                    branch[r][c] = slash
                else:
                    if slash == "/":
                        branch[r][c + 1] = slash
                    else:
                        branch[r][c - 1] = slash
                if cur.left:
                    bq.append([r + 1, c - 2 ** (m - r - 2), cur.left, "/"])
                if cur.right:
                    bq.append([r + 1, c + 2 ** (m - r - 2), cur.right, "\\"])

        for i in range(m):
            print("".join(branch[i]))
            print("".join(ans[i]))


if __name__ == "__main__":
    tree = BinaryTree()
    # for i in [63909, 43838, 4549, -31714, -99701, -96320, 88666, 75152,
    #           -14750, -12671, 60405, None, 29388, None, None, None,
    #           -83674, None, None, -83490, None, -49913, 86188, 84455,
    #           None, None, None, None, None, None, None, None, None,
    #           -36061, 91438, -75550]:
    #     tree.add(i)
    for i in [1, None, 2, None, 3, 4, 5]:
        tree.add(i)
    # for i in range(25):
    #     tree.add(chr(97 + i))

    tree.print_Tree()
    # for i in range(1, 8):
    # tree.add(i)
    # print("广度遍历0: ")
    # tree.breadth_travel()
    print("广度遍历1: ")
    tree.breadth_travel1()
    print("广度遍历2: ")
    tree.breadth_travel2()

    print("\n深度遍历: ")
    print("前序遍历: 递归")
    tree.pre_order()
    # print("前序遍历: 非递归")
    # tree.pre_order_1()

    print()
    print("中序遍历: 递归")
    tree.in_order()
    # print("中序遍历: 非递归, 用指针")
    # tree.in_order_1()
    # print("中序遍历: 非递归, 不需要指针")
    # tree.in_order_2()
    # print("中序遍历: 非递归, 颜色标记法")
    # tree.in_order_3()
    # print("中序遍历: 非递归, 颜色标记法(改进)")
    # tree.in_order_4()

    print()
    print("后序遍历: 递归")
    tree.post_order()
    # print("后序遍历: 非递归, 修改自前序")
    # tree.post_order_1()
    # print("后序遍历: 非递归, 直接写")
    # tree.post_order_2()

"""
广度遍历: 
0 1 2 3 4 5 6 7 8 
深度遍历: 
前序遍历: 递归
0 1 3 7 8 4 2 5 6 
中序遍历: 递归
7 3 8 1 4 0 5 2 6 
后序遍历: 递归
7 8 3 4 1 5 6 2 0 
前序遍历: 非递归
[0, 1, 3, 7, 8, 4, 2, 5, 6]
中序遍历: 非递归
[7, 3, 8, 1, 4, 0, 5, 2, 6]
后序遍历: 非递归
[7, 8, 3, 4, 1, 5, 6, 2, 0]
"""
